### PR TITLE
[JAVA CLIENT] Add Support to Override Headers for each Request

### PR DESCRIPTION
**Pull request by cyrusv**
My first contribution -- hoping everything is set up ok! 

Resolves https://github.com/OpenAPITools/openapi-generator/issues/9792. Basically, I want to use a shared connection, but I want to specify different JWT for each request. This has to be done by overloading the calls (some calls have no params or primitive params, so I need to have a new param). I implemented as a headers map to be more generally useful.

I added a  `Map<String, String> headers` to the client calls, and passed it down the stack.

While I was fixing, I found it difficult to reason about when these headers were being modified, because the `String[] authNames` parameter was kind of goofy, and all the request headers were not modified in the same place -- I changed the code to now do all the header changes/mutations in one place. This makes it easier to know for certain that any headers intended to override are in fact added after all other mutations to the headers.

Hoping for feedback from Java reviewers:
@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog  @lwlee2608  @nmuesch 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Also I'm an openapi newbie, so would certainly appreciate some help understanding the test failures in CI?

### CODE REVIEW

**Comment by cyrusv** [comment]
@MosheElisha  or @wing328  -- any help here?  I will fix the tests, but I want to be sure the strategy is generally acceptable before I sink more time into it.

**Comment by wing328** [comment]
@cyrusv thanks for the PR. I'll try to review tomorrow or later this week and let you know.

**Comment by MosheElisha** [comment]
@cyrusv Thanks for the PR.

The approach I take when I need to control headers per request is to use ThreadLocal and interceptor.
When using the client you do something like:

```
try (RequestModifier requestModifier = new RequestModifier()
        .addHeader("Authorization", authToken)) {

  myClient.deleteBook(bookId);

}
```
RequestModifier stores the headers in a ThreadLocal map and an interceptor reads those headers from the thread and adds them to the request.
[RequestModifier.zip](https://github.com/OpenAPITools/openapi-generator/files/6790255/RequestModifier.zip)

I have added RequestModifier code. The implementation of the RequestModifier and interceptor implementation is different for every `library` but it is pretty straight forward.

File: modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/api.mustache [hunk 1]
```diff
@@ -96,7 +96,7 @@ public class {{classname}} {
     {{#isDeprecated}}
     @Deprecated
     {{/isDeprecated}}
-    public{{/vendorExtensions.x-group-parameters}}{{#vendorExtensions.x-group-parameters}}private{{/vendorExtensions.x-group-parameters}} okhttp3.Call {{operationId}}Call({{#allParams}}{{{dataType}}} {{paramName}}, {{/allParams}}final ApiCallback _callback) throws ApiException {
+    public{{/vendorExtensions.x-group-parameters}}{{#vendorExtensions.x-group-parameters}}private{{/vendorExtensions.x-group-parameters}} okhttp3.Call {{operationId}}Call({{#allParams}}{{{dataType}}} {{paramName}}, {{/allParams}}Map<String, String> headers, final ApiCallback _callback) throws ApiException {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/api.mustache:99]: > Map<String, String> headers, 

My opinion is that I prefer not to add an additional parameter for every methods.

Have you considered using the default headers in the ApiClient object to achieve what you want?  

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/ApiClient.mustache#L529

Comment (cyrusv) [reply modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/api.mustache:99]: Thanks, @wing328  -- I had considered this, but that wouldn't be a threadsafe way to do it in a case where multiple threads are sharing the instance.

**Comment by cyrusv** [comment]
Thanks @MosheElisha, I didn't know about Request Modifiers. Let me see if they can solve my problem.

**Pull request closed without merging**